File: agent/library/src/main/java/org/matilda/commands/listener/CommandResponseListener.java
package org.matilda.commands.listener;


import com.google.protobuf.InvalidProtocolBufferException;
import org.matilda.commands.protobuf.CommandResponse;
import org.matilda.messages.listener.MessageListener;
import org.matilda.messages.listener.MessageListeningInstance;
import org.matilda.messages.protobuf.MessageType;

import javax.inject.Inject;

public class CommandResponseListener {
    @Inject
    MessageListener mMessageListener;

    @Inject
    public CommandResponseListener() {}

    public CommandResponseListeningInstance listen(int commandId) {
        MessageListeningInstance listeningInstance = mMessageListener.listen(MessageType.COMMAND_RESPONSE_VALUE, message -> {
            try {
                return CommandResponse.parseFrom(message.data).getId() == commandId;
            } catch (InvalidProtocolBufferException e) {
                throw new RuntimeException(e);
            }
        });

        return new CommandResponseListeningInstance(listeningInstance);
    }
}

File: agent/library/src/main/java/org/matilda/commands/listener/CommandResponseListener.kt
package org.matilda.commands.listener

import me.tatarka.inject.annotations.Inject
import org.matilda.commands.protobuf.CommandResponse
import org.matilda.messages.listener.MessageListener
import org.matilda.messages.protobuf.MessageType

class CommandResponseListener @Inject constructor(private val mMessageListener: MessageListener) {
    fun listen(commandId: Int): CommandResponseListeningInstance {
        val listeningInstance = mMessageListener.listen(MessageType.COMMAND_RESPONSE_VALUE) {
            CommandResponse.parseFrom(it.data).id == commandId
        }
        return CommandResponseListeningInstance(listeningInstance)
    }
}
